src/Monotonic.cpp

```diff
@@ -30,26 +30,173 @@ using std::string;
+    return a.has_upper_bound() && a.max <= 0;
+}
+
+ConstantInterval to_interval(Monotonic m) {
+    switch (m) {
+    case Monotonic::Constant:
+        return ConstantInterval::single_point(0);
+    case Monotonic::Increasing:
+        return ConstantInterval::bounded_below(0);
+    case Monotonic::Decreasing:
+        return ConstantInterval::bounded_above(0);
+    case Monotonic::Unknown:
+        return ConstantInterval::everything();
+    }
+    return ConstantInterval::everything();
+}
+
+Monotonic to_monotonic(const ConstantInterval &x) {
+    if (is_constant(x)) {
+        return Monotonic::Constant;
+    } else if (is_monotonic_increasing(x)) {
+        return Monotonic::Increasing;
+    } else if (is_monotonic_decreasing(x)) {
+        return Monotonic::Decreasing;
+    } else {
+        return Monotonic::Unknown;
+    }
+}
+
+ConstantInterval unify(const ConstantInterval &a, const ConstantInterval &b) {
+    return ConstantInterval::make_union(a, b);
+}
+
+ConstantInterval unify(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result;
+    result.include(b);
+    return result;
+}
+
+// Helpers for doing arithmetic on ConstantIntervals that avoid generating
+// expressions of pos_inf/neg_inf.
+ConstantInterval add(const ConstantInterval &a, const ConstantInterval &b) {
+    ConstantInterval result;
+    result.min_defined = a.has_lower_bound() && b.has_lower_bound();
+    result.max_defined = a.has_upper_bound() && b.has_upper_bound();
+    if (result.has_lower_bound()) {
+        result.min = a.min + b.min;
+    }
+    if (result.has_upper_bound()) {
+        result.max = a.max + b.max;
+    }
+    return result;
+}
+
+ConstantInterval add(const ConstantInterval &a, int64_t b) {
+    return add(a, ConstantInterval(b, b));
+}
+
+ConstantInterval negate(const ConstantInterval &r) {
+    ConstantInterval result;
+    result.min_defined = r.has_upper_bound();
+    result.min = r.has_upper_bound() ? -r.max : 0;
+    result.max_defined = r.has_lower_bound();
+    result.max = r.has_lower_bound() ? -r.min : 0;
+    return result;
+}
+
+ConstantInterval sub(const ConstantInterval &a, const ConstantInterval &b) {
+    return add(a, negate(b));
+}
+
+ConstantInterval sub(const ConstantInterval &a, int64_t b) {
+    return sub(a, ConstantInterval(b, b));
+}
+
+ConstantInterval multiply(const ConstantInterval &a, int64_t b) {
+    ConstantInterval result(a);
+    if (b < 0) {
+        result = negate(result);
+        b = -b;
+    }
+    if (result.has_lower_bound()) {
+        result.min *= b;
+    }
+    if (result.has_upper_bound()) {
+        result.max *= b;
+    }
+    return result;
+}
+
+ConstantInterval multiply(const ConstantInterval &a, const ConstantInterval &b) {
+    int64_t bounds[4];
+    int64_t *bounds_begin = &bounds[0];
+    int64_t *bounds_end = &bounds[0];
+    ConstantInterval result;
+    result.min_defined = result.max_defined = true;
+    if (a.has_lower_bound() && b.has_lower_bound()) {
```

Comment:
This series of rules seems more confusing than it needs to be - the result never can produce a partially bounded value.

i.e. for `result.min_defined` to be true, we need `a.has_lower_bound() && b.has_upper_bound()` and `a.has_upper_bound() && b.has_lower_bound()`, and similarly `a.has_lower_bound() && b.has_lower_bound()` and `a.has_upper_bound() && b.has_upper_bound()` for `result.max_defined` to be true. At this point, if `a` and `b` are both bounded, then compare all 4 values and choose the mins/maxs for the same result.

Reply:
Agreed this was overly complicated. I think with the refactored representation, it's much easier.

Are you sure that the result is never a partially bounded interval? What about `[1, inf) * [0, 1]` ? Shouldn't this be `[0, inf)` ?

Reply:
Oh sorry, I should have said that better. The result of a multiply could absolutely be a partially bounded interval, like your example - but I don't think that this function ever produces a partially bounded interval.

Reply:
If I run:
```
    ConstantInterval a = ConstantInterval(0, 1);
    ConstantInterval b = ConstantInterval::bounded_below(1);
    ConstantInterval c = multiply(a, b);
```
On this branch, I produce an interval that has undefined min and max

Reply:
Oh oops, I see there was a bit of a refactor, that might no longer be the case.

Reply:
Now this example no longer works because:
```
ConstantInterval::bounded_below(1);
```
Throws an internal assert, because `bounded_below` seems to assume a negative argument, is this true?

Reply:
Ok so if I make `bounded_below` use `std::numeric_limits<int64_t>::max()` in the ctor to avoid that error, this example still produces a `c` that has undefined min and max.

Reply:
The example now works, and I *think* the casework is correct. I haven't found a prettier way to avoid the series of cases yet either..